// @colony/tokens/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenClientType = exports.ClientType = void 0;
/**
 * The type for a specific contract-client (extended ethers `Contract`).
 * This is being used for TypeScript's discriminative unions (to make assumptions about what functionality is available on this contract)
 */
var ClientType;
(function (ClientType) {
    ClientType["TokenClient"] = "TokenClient";
    ClientType["TokenLockingClient"] = "TokenLockingClient";
})(ClientType || (exports.ClientType = ClientType = {}));
/**
 * We support different TokenClients, especially the ColonyToken client with
 * its advanced functionality (to `.mint()` tokens for example). Other tokens
 * require certain adjustments (like the original Dai (SAI))
 */
var TokenClientType;
(function (TokenClientType) {
    TokenClientType["Colony"] = "Colony";
    TokenClientType["ColonyLegacy"] = "ColonyLegacy";
    TokenClientType["Erc20"] = "Erc20";
    TokenClientType["Erc2612"] = "Erc2612";
    TokenClientType["Sai"] = "Sai";
})(TokenClientType || (exports.TokenClientType = TokenClientType = {}));
