// .ignored/@colony/sdk/dist/cjs/ColonyNetwork/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vote = exports.VotingReputation = exports.TokenLocking = exports.OneTxPayment = exports.SupportedExtension = exports.Colony = exports.ColonyNetwork = void 0;
var ColonyNetwork_js_1 = require("./ColonyNetwork.js");
Object.defineProperty(exports, "ColonyNetwork", { enumerable: true, get: function () { return ColonyNetwork_js_1.ColonyNetwork; } });
var Colony_js_1 = require("./Colony.js");
Object.defineProperty(exports, "Colony", { enumerable: true, get: function () { return Colony_js_1.Colony; } });
Object.defineProperty(exports, "SupportedExtension", { enumerable: true, get: function () { return Colony_js_1.SupportedExtension; } });
var OneTxPayment_js_1 = require("./OneTxPayment.js");
Object.defineProperty(exports, "OneTxPayment", { enumerable: true, get: function () { return OneTxPayment_js_1.OneTxPayment; } });
var TokenLocking_js_1 = require("./TokenLocking.js");
Object.defineProperty(exports, "TokenLocking", { enumerable: true, get: function () { return TokenLocking_js_1.TokenLocking; } });
var VotingReputation_js_1 = require("./VotingReputation.js");
Object.defineProperty(exports, "VotingReputation", { enumerable: true, get: function () { return VotingReputation_js_1.VotingReputation; } });
Object.defineProperty(exports, "Vote", { enumerable: true, get: function () { return VotingReputation_js_1.Vote; } });
__exportStar(require("./tokens/index.js"), exports);
__exportStar(require("./CustomContract.js"), exports);
