// @colony/tokens@1.1.0/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenClientType = exports.ClientType = exports.getTokenLockingClient = exports.getTokenClient = exports.ERC2612TokenFactory = exports.ERC20TokenFactory = exports.ColonyTokenFactory = exports.ERC2612 = exports.ERC721 = exports.ERC20 = void 0;
const ethers_1 = require("ethers");
const index_js_1 = require("./contracts/static/index.js");
const { AddressZero } = ethers_1.constants;
// Initialize dummy ethers contracts in order to get their (typed) interfaces
exports.ERC20 = index_js_1.TokenERC20__factory.connect(AddressZero, new ethers_1.providers.BaseProvider(3656691)).interface;
exports.ERC721 = index_js_1.TokenERC721__factory.connect(AddressZero, new ethers_1.providers.BaseProvider(3656691)).interface;
exports.ERC2612 = index_js_1.TokenERC2612__factory.connect(AddressZero, new ethers_1.providers.BaseProvider(3656691)).interface;
// Some nice aliases for the important token factories
var index_js_2 = require("./contracts/index.js");
Object.defineProperty(exports, "ColonyTokenFactory", { enumerable: true, get: function () { return index_js_2.MetaTxToken__factory; } });
var index_js_3 = require("./contracts/static/index.js");
Object.defineProperty(exports, "ERC20TokenFactory", { enumerable: true, get: function () { return index_js_3.TokenERC20__factory; } });
var index_js_4 = require("./contracts/static/index.js");
Object.defineProperty(exports, "ERC2612TokenFactory", { enumerable: true, get: function () { return index_js_4.TokenERC2612__factory; } });
__exportStar(require("./contracts/index.js"), exports);
var TokenClient_js_1 = require("./clients/TokenClient.js");
Object.defineProperty(exports, "getTokenClient", { enumerable: true, get: function () { return __importDefault(TokenClient_js_1).default; } });
var TokenLockingClient_js_1 = require("./clients/TokenLockingClient.js");
Object.defineProperty(exports, "getTokenLockingClient", { enumerable: true, get: function () { return __importDefault(TokenLockingClient_js_1).default; } });
var constants_js_1 = require("./constants.js");
Object.defineProperty(exports, "ClientType", { enumerable: true, get: function () { return constants_js_1.ClientType; } });
Object.defineProperty(exports, "TokenClientType", { enumerable: true, get: function () { return constants_js_1.TokenClientType; } });
