// @colony/core@3.1.0/dist/cjs/contracts/factories/MotionTarget__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MotionTarget__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        inputs: [
            {
                internalType: "bytes4",
                name: "_sig",
                type: "bytes4",
            },
        ],
        name: "getCapabilityRoles",
        outputs: [
            {
                internalType: "bytes32",
                name: "roles",
                type: "bytes32",
            },
        ],
        stateMutability: "view",
        type: "function",
    },
];
class MotionTarget__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.MotionTarget__factory = MotionTarget__factory;
MotionTarget__factory.abi = _abi;
