// @colony/events@4.1.0/dist/cjs/contracts.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WhitelistEvents = exports.VotingReputationEvents = exports.TokenSupplierEvents = exports.TokenLockingEvents = exports.TokenEvents = exports.StreamingPaymentsEvents = exports.StakedExpenditureEvents = exports.StagedExpenditureEvents = exports.ReputationBootstrapperEvents = exports.OneTxPaymentEvents = exports.MultisigPermissionsEvents = exports.MetaTxTokenEvents = exports.FundingQueueEvents = exports.EvaluatedExpenditureEvents = exports.ColonyNetworkEvents = exports.ColonyExtensionEvents = exports.ColonyEvents = exports.CoinMachineEvents = void 0;
exports.CoinMachineEvents = __importStar(require("./contracts/CoinMachineEvents.js"));
exports.ColonyEvents = __importStar(require("./contracts/IColonyEvents.js"));
exports.ColonyExtensionEvents = __importStar(require("./contracts/ColonyExtensionEvents.js"));
exports.ColonyNetworkEvents = __importStar(require("./contracts/IColonyNetworkEvents.js"));
exports.EvaluatedExpenditureEvents = __importStar(require("./contracts/EvaluatedExpenditureEvents.js"));
exports.FundingQueueEvents = __importStar(require("./contracts/FundingQueueEvents.js"));
exports.MetaTxTokenEvents = __importStar(require("./contracts/MetaTxTokenEvents.js"));
exports.MultisigPermissionsEvents = __importStar(require("./contracts/MultisigPermissionsEvents.js"));
exports.OneTxPaymentEvents = __importStar(require("./contracts/OneTxPaymentEvents.js"));
exports.ReputationBootstrapperEvents = __importStar(require("./contracts/ReputationBootstrapperEvents.js"));
exports.StagedExpenditureEvents = __importStar(require("./contracts/StagedExpenditureEvents.js"));
exports.StakedExpenditureEvents = __importStar(require("./contracts/StakedExpenditureEvents.js"));
exports.StreamingPaymentsEvents = __importStar(require("./contracts/StreamingPaymentsEvents.js"));
exports.TokenEvents = __importStar(require("./contracts/TokenEvents.js"));
exports.TokenLockingEvents = __importStar(require("./contracts/TokenLockingEvents.js"));
exports.TokenSupplierEvents = __importStar(require("./contracts/TokenSupplierEvents.js"));
exports.VotingReputationEvents = __importStar(require("./contracts/IVotingReputationEvents.js"));
exports.WhitelistEvents = __importStar(require("./contracts/WhitelistEvents.js"));
__exportStar(require("./contracts/factories/index.js"), exports);
