// @colony/events@4.1.0/dist/esm/ipfs/constants.js
/**
 * Available metadata types
 *
 * DEFAULT is usually used to indicate a generic Metadata type
 */
export var MetadataType;
(function (MetadataType) {
    MetadataType["DEFAULT"] = "default";
    MetadataType["Annotation"] = "annotation";
    MetadataType["Colony"] = "colony";
    MetadataType["Decision"] = "decision";
    MetadataType["Domain"] = "domain";
    MetadataType["Misc"] = "misc";
})(MetadataType || (MetadataType = {}));
/**
 * Current Colony Event Metadata version
 */
export const METADATA_VERSION = 2;
/**
 * The various team colors that the dApp supports
 */
export var TeamColor;
(function (TeamColor) {
    TeamColor[TeamColor["LightPink"] = 0] = "LightPink";
    TeamColor[TeamColor["Pink"] = 1] = "Pink";
    TeamColor[TeamColor["Black"] = 2] = "Black";
    TeamColor[TeamColor["EmeraldGreen"] = 3] = "EmeraldGreen";
    TeamColor[TeamColor["Blue"] = 4] = "Blue";
    TeamColor[TeamColor["Yellow"] = 5] = "Yellow";
    TeamColor[TeamColor["Red"] = 6] = "Red";
    TeamColor[TeamColor["Green"] = 7] = "Green";
    TeamColor[TeamColor["Periwinkle"] = 8] = "Periwinkle";
    TeamColor[TeamColor["Gold"] = 9] = "Gold";
    TeamColor[TeamColor["Aqua"] = 10] = "Aqua";
    TeamColor[TeamColor["BlueGrey"] = 11] = "BlueGrey";
    TeamColor[TeamColor["Purple"] = 12] = "Purple";
    TeamColor[TeamColor["Orange"] = 13] = "Orange";
    TeamColor[TeamColor["Magenta"] = 14] = "Magenta";
    TeamColor[TeamColor["PurpleGrey"] = 15] = "PurpleGrey";
})(TeamColor || (TeamColor = {}));
