// @colony/events@4.1.0/dist/cjs/contracts/factories/ReputationBootstrapperEvents__factory.js
"use strict";
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ReputationBootstrapperEvents__factory = void 0;
const ethers_1 = require("ethers");
const _abi = [
    {
        anonymous: false,
        inputs: [],
        name: "ExtensionInitialised",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "recipient",
                type: "address",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "reputationAmount",
                type: "uint256",
            },
            {
                indexed: false,
                internalType: "bool",
                name: "paid",
                type: "bool",
            },
        ],
        name: "GrantClaimed",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "bool",
                name: "paid",
                type: "bool",
            },
            {
                indexed: false,
                internalType: "bytes32",
                name: "hashedSecret",
                type: "bytes32",
            },
            {
                indexed: false,
                internalType: "uint256",
                name: "reputationAmount",
                type: "uint256",
            },
        ],
        name: "GrantSet",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "authority",
                type: "address",
            },
        ],
        name: "LogSetAuthority",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "owner",
                type: "address",
            },
        ],
        name: "LogSetOwner",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "user",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address payable",
                name: "relayerAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "functionSignature",
                type: "bytes",
            },
        ],
        name: "MetaTransactionExecuted",
        type: "event",
    },
];
class ReputationBootstrapperEvents__factory {
    static createInterface() {
        return new ethers_1.utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new ethers_1.Contract(address, _abi, signerOrProvider);
    }
}
exports.ReputationBootstrapperEvents__factory = ReputationBootstrapperEvents__factory;
ReputationBootstrapperEvents__factory.abi = _abi;
