// .ignored/@colony/sdk/dist/cjs/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContractConfig = exports.ERC721Interface = exports.ERC2612Interface = exports.ERC20Interface = exports.w = exports.toWei = exports.toEth = exports.parseLogs = exports.isExtensionCompatible = exports.addressesAreEqual = exports.Tokens = exports.Network = exports.MotionState = exports.Id = exports.ExtensionVersions = exports.Extension = exports.ColonyRpcEndpoint = exports.ColonyRole = exports.TeamColor = exports.MetadataType = void 0;
var events_1 = require("@colony/events");
Object.defineProperty(exports, "MetadataType", { enumerable: true, get: function () { return events_1.MetadataType; } });
Object.defineProperty(exports, "TeamColor", { enumerable: true, get: function () { return events_1.TeamColor; } });
var core_1 = require("@colony/core");
Object.defineProperty(exports, "ColonyRole", { enumerable: true, get: function () { return core_1.ColonyRole; } });
Object.defineProperty(exports, "ColonyRpcEndpoint", { enumerable: true, get: function () { return core_1.ColonyRpcEndpoint; } });
Object.defineProperty(exports, "Extension", { enumerable: true, get: function () { return core_1.Extension; } });
Object.defineProperty(exports, "ExtensionVersions", { enumerable: true, get: function () { return core_1.ExtensionVersions; } });
Object.defineProperty(exports, "Id", { enumerable: true, get: function () { return core_1.Id; } });
Object.defineProperty(exports, "MotionState", { enumerable: true, get: function () { return core_1.MotionState; } });
Object.defineProperty(exports, "Network", { enumerable: true, get: function () { return core_1.Network; } });
Object.defineProperty(exports, "Tokens", { enumerable: true, get: function () { return core_1.Tokens; } });
Object.defineProperty(exports, "addressesAreEqual", { enumerable: true, get: function () { return core_1.addressesAreEqual; } });
Object.defineProperty(exports, "isExtensionCompatible", { enumerable: true, get: function () { return core_1.isExtensionCompatible; } });
Object.defineProperty(exports, "parseLogs", { enumerable: true, get: function () { return core_1.parseLogs; } });
Object.defineProperty(exports, "toEth", { enumerable: true, get: function () { return core_1.toEth; } });
Object.defineProperty(exports, "toWei", { enumerable: true, get: function () { return core_1.toWei; } });
Object.defineProperty(exports, "w", { enumerable: true, get: function () { return core_1.w; } });
__exportStar(require("@colony/events"), exports);
var tokens_1 = require("@colony/tokens");
Object.defineProperty(exports, "ERC20Interface", { enumerable: true, get: function () { return tokens_1.ERC20; } });
Object.defineProperty(exports, "ERC2612Interface", { enumerable: true, get: function () { return tokens_1.ERC2612; } });
Object.defineProperty(exports, "ERC721Interface", { enumerable: true, get: function () { return tokens_1.ERC721; } });
__exportStar(require("./ColonyNetwork/index.js"), exports);
var ContractConfig_js_1 = require("./ContractConfig.js");
Object.defineProperty(exports, "ContractConfig", { enumerable: true, get: function () { return ContractConfig_js_1.ContractConfig; } });
__exportStar(require("./TxCreator/index.js"), exports);
