// @colony/events@4.1.0/dist/esm/contracts/factories/TokenSupplierEvents__factory.js
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import { Contract, utils } from "ethers";
const _abi = [
    {
        anonymous: false,
        inputs: [],
        name: "ExtensionInitialised",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "authority",
                type: "address",
            },
        ],
        name: "LogSetAuthority",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: true,
                internalType: "address",
                name: "owner",
                type: "address",
            },
        ],
        name: "LogSetOwner",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "tokenIssuanceRate",
                type: "uint256",
            },
        ],
        name: "TokenIssuanceRateSet",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "tokenSupplyCeiling",
                type: "uint256",
            },
        ],
        name: "TokenSupplyCeilingSet",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "uint256",
                name: "numTokens",
                type: "uint256",
            },
        ],
        name: "TokensIssued",
        type: "event",
    },
    {
        anonymous: false,
        inputs: [
            {
                indexed: false,
                internalType: "address",
                name: "user",
                type: "address",
            },
            {
                indexed: false,
                internalType: "address payable",
                name: "relayerAddress",
                type: "address",
            },
            {
                indexed: false,
                internalType: "bytes",
                name: "functionSignature",
                type: "bytes",
            },
        ],
        name: "MetaTransactionExecuted",
        type: "event",
    },
];
export class TokenSupplierEvents__factory {
    static abi = _abi;
    static createInterface() {
        return new utils.Interface(_abi);
    }
    static connect(address, signerOrProvider) {
        return new Contract(address, _abi, signerOrProvider);
    }
}
